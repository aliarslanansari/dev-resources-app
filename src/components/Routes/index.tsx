import React from 'react'
import { useSelector } from 'react-redux'
import { Redirect, Route, Switch } from 'react-router-dom'
import { selectIsLoggedIn } from '../../container/LoginPageContainer/selectors'
import { RouteConfig } from '../../types'

interface RoutePropsTypes {
  routeConfig: RouteConfig
}

const Routes = (props: RoutePropsTypes) => {
  const { routeConfig } = props
  const isLoggedIn = useSelector(selectIsLoggedIn())
  return (
    <Switch>
      {Object.keys(routeConfig).map((routeKey, index) => {
        if (routeConfig[routeKey].isProtected && !isLoggedIn) {
          return (
            <Redirect
              to={`${routeConfig.login.path}?redirectURL=${routeConfig[routeKey].path}`}
            />
          )
        } else {
          return (
            <Route
              exact={routeConfig[routeKey].exact}
              key={index}
              path={routeConfig[routeKey].path}
              render={routeConfig[routeKey].render}
            />
          )
        }
      })}
    </Switch>
  )
}

export default Routes
