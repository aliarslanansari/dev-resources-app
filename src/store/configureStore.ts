import { createStore, applyMiddleware, compose } from 'redux'
import { createInjectorsEnhancer } from 'redux-injectors'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import immutableTransform from 'redux-persist-transform-immutable'
import { BrowserHistory } from '../utils/history'
import createSagaMiddleware from 'redux-saga'
import { routerMiddleware } from 'connected-react-router'
import createReducer from './rootReducer'

const persistConfig = {
  transforms: [immutableTransform()],
  key: 'root',
  blacklist: ['router'],
  storage
}

const persistedReducer = persistReducer(persistConfig, createReducer())

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose
  }
}

const configureStore = (initialState: Object, history: BrowserHistory) => {
  let composeEnhancers = compose
  const reduxSagaMonitorOptions = {}
  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) {
      //@ts-ignore
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    }
  }
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions)
  const runSaga = sagaMiddleware.run
  const injectEnhancer = createInjectorsEnhancer({
    createReducer,
    runSaga
  })

  const middlewares = [sagaMiddleware, routerMiddleware(history)]

  const enhancers = [applyMiddleware(...middlewares), injectEnhancer]

  const store = createStore(
    persistedReducer,
    initialState,
    composeEnhancers(...enhancers)
  )

  let persistor = persistStore(store)
  return { store, persistor }
}

export default configureStore
