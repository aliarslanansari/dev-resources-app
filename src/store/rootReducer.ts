import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import history from '../utils/history'
import loginPageContainer from '../container/LoginPageContainer/reducer'
import registerUserContainerReducer from '../container/RegisterUserContainer/reducer'
import homeContainerReducer from '../container/HomeContainer/reducer'
import createPostContainer from '../container/CreatePostContainer/reducer'
import postDetailsContainer from '../container/PostDetails/reducer'

export default function createReducer() {
  const appReducer = combineReducers({
    router: connectRouter(history),
    loginPageContainer,
    registerUserContainerReducer,
    homeContainerReducer,
    createPostContainer,
    postDetailsContainer
  })

  const rootReducer = (state: any, action: any) => {
    // if (
    //   action.type === loginPageContainerTypes.SUCCESS_USER_LOGOUT ||
    //   action.type === loginPageContainerTypes.FAILURE_USER_LOGOUT
    // ) {
    //   state = undefined
    // }
    return appReducer(state, action)
  }
  return rootReducer
}
