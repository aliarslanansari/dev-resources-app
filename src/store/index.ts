import history from '../utils/history'
import configureStore from './configureStore'
const initialState = {}
export const { store, persistor } = configureStore(initialState, history)

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch

export default store
