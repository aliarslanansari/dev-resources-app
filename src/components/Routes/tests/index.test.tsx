import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import store from '../../../store'
import { RouteConfig } from '../../../types'
import Routes from '../index'

const routeConfig: RouteConfig = {
  login: {
    path: '/login',
    render: () => <h1 data-testid='login-test'>Login</h1>,
    exact: true
  },
  createPost: {
    path: '/createPost',
    render: () => <h1 data-testid='createPost-test'>createPost</h1>,
    exact: true,
    isProtected: true
  }
}

describe('<DRButton />', () => {
  it('should render and match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/createPost']}>
          <Routes isLoggedIn={true} routeConfig={routeConfig} />
        </MemoryRouter>
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should redirect to login page if not logged in', () => {
    const { getAllByTestId } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/createPost']}>
          <Routes isLoggedIn={false} routeConfig={routeConfig} />
        </MemoryRouter>
      </Provider>
    )
    expect(getAllByTestId('login-test').length).toBe(1)
  })
  it('should allow to access page page if logged in', () => {
    const { getAllByTestId } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/createPost']}>
          <Routes isLoggedIn={true} routeConfig={routeConfig} />
        </MemoryRouter>
      </Provider>
    )
    expect(getAllByTestId('createPost-test').length).toBe(1)
  })
})
