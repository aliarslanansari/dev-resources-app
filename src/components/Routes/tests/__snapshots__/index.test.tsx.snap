// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DRButton /> should render and match the snapshot 1`] = `
<div>
  <h1
    data-testid="createPost-test"
  >
    createPost
  </h1>
</div>
`;
